Compatibility-2.0.md

```diff
@@ -26,3 +26,304 @@ Embedded Mode also might use a decorated Cache using a Transcoder. For example t
+
+## About Encoders
+
+An Encoder is essentially a two-way conversion between the storage format in a cache and the read/write format. The ```AdvancedCache```  interface has a method to specify an encoder to be used during all operations involving that cache:
+
+```cache.getAdvancedCache().withEncoding(MyEncoder.class)```
+
+which returns a decorated cache instance where all operations (read/write/stream/add listener, etc) flow through the supplied encoder.
+
+At the moment, there is no cache configuration to specify an Encoder; Infinispan tries to pick the best decorated cache depending on the
+situation where this cache is used. For example:
+
+* Embedded caches with ```Storage.OBJECT``` are associated with the ```IdentityEncoder```, which stores objects "as is";
+* Embedded caches with ```Storage.OFF_HEAP``` are associated with a ```MarshallerEncoder```, which marshalls data before storing in order to produce a ```byte[]``` (hard requirement);
+* Embedded caches with ```Storage.BINARY``` are associated with the BinaryEncoder, which produces a mix of marshalled and unmarshalled objects.
+* Caches used in the server with indexing enabled and compatibility mode enabled uses a ```ProtostreamCompatEncoder```, where the byte[] arriving from the client
+ is unmarshalled via a protostream marshaller into a java representation.
+* Caches used in the server with no indexing and compatibility use the ```IdentityEncoder```, which treats content read and written as opaque ```byte[]```.
+* Caches used in the server with compatibility (and no indexing) uses the CompatModeEncoder, which takes the ```byte[]``` from the client and unmarshalls them to java object for storage.
+* Scripts which carry data-type ```utf-8``` wraps a cache with the ```UTF8Encoder```, that will extract ```byte[]``` out of a String in order to store and will create Strings out of ```bytes[]``` for reading
+operations.
+
+
+Some endpoints in the Infinispan Server allow to send and request data in a certain format (MimeType):
+
+* Scripting: scripts (.js at least) can pass a header "dataType" with a MimeType string;
+* Rest: request headers "content-type" (for PUT/POST) and "accept" (for GET) can carry mime types as well
+* HotRod: there is no such capability yet, but it's something that will eventually be added.
+
+
+How does the Mimetype relate to the encoder general mechanism?
+
+
+## Proposal
+
+The proposal is to use a pair (Encoder, MimeType) to describe cache storage. Examples: 
+
+* "```application/json``` with no encoding" means IdentityEncoder associated of a content that represents a json (```byte[]```, ```Gson```, ```Map<String, Map<?,?>>```, etc)
+* "```text/plain; charset=utf-8``` with marshalling" means content produced by the MarshallerEncoder that represents text, so it's a ```byte[]```
+* "```application/octet-stream``` with KryoEncoder" means ```byte[]``` produced by a KryoMarshaller.
```

Comment:
I found it just recently in REST, sometimes users don't care about the format. They just use `Content-type: */*`. I think we should consider doing a fallback `text/plain` for example. 

Also you may be interested in making charset optional or even "replaceable". E.g. one might store entry in UTF8 but get it using ASCII. The good news is that the charset problem touches only "text" mime-types. For vast majority things like JSON, `UTF-8` is the only valid encoding.

The point I'm trying to make is that it *may* be n to 1 mapping (n mime-types to 1 converter). Just make sure we support this well enough.